### ISSUE TITLE
[CD] Deploy generated app in Heroku

### ISSUE THREAD

**Issue by nicolas63** [jhipster/jhipster-dotnetcore#389]
##### **Overview of the feature request**

Deploy generated app in Heroku

##### **Related issues or PR**

-   [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by ivanmonteiro** [comment]
Would this issue add the required configuration files to run the generated app on heroku? Or actually publish the generated project there?
Any reference points for this issue?

**Comment by nicolas63** [comment]
You just need custom build pack like https://medium.com/ingenious/net-core-and-heroku-74e6b41e8c64

**Comment by nicolas63** [comment]
@ivanmonteiro you closed this issue a bit quickly I didn't have time to read your PR because your PR was in draft. i have some remarks why disable okta (we support okta with open-id authentication ) ? Can you clean the commented code ? The blazor part support this deployment ? if not i can help you to add it ?

**Comment by ivanmonteiro** [comment]
> @ivanmonteiro you closed this issue a bit quickly I didn't have time to read your PR because your PR was in draft. i have some remarks why disable okta (we support okta with open-id authentication ) ? Can you clean the commented code ? The blazor part support this deployment ? if not i can help you to add it ?

I have explained on gitter what happened. Summing up:
I mistakenly pushed my work-in-progress to upstream instead of origin.
There is still work to be done: okta, clean commented code, test blazor deploy, etc... And of course a proper PR review.
I need some guidance on how to proceed now.

**Comment by ivanmonteiro** [comment]
I have all the code on my fork feature branch and on my machine, so if main branch It needs to be reverted I have everything here.

**Comment by nicolas63** [comment]
@ivanmonteiro oh ok, we can discuss it on gitter

**Comment by ivanmonteiro** [comment]
Continuing work on a new PR: https://github.com/jhipster/jhipster-dotnetcore/pull/561

**Comment by ivanmonteiro** [comment]
@nicolas63 I'm having some errors trying to deploy the app with blazor.

First, I needed to install the tools needed as dependencies by blazor: libman and webcompiler.  

I managed to do that by createing a new tools manifest (.config/dotnet-tools.json) with:
 `dotnet new tool-manifest` and adding the tools needed:
`dotnet tool install -g Microsoft.Web.LibraryManager.Cli`
and 
`dotnet tool install Excubo.WebCompiler`.

This creates .config/dotnet-tools.json file with the following contents:
```
{
  "version": 1,
  "isRoot": true,
  "tools": {
    "microsoft.web.librarymanager.cli": {
      "version": "2.1.113",
      "commands": [
        "libman"
      ]
    },
    "excubo.webcompiler": {
      "version": "2.6.3",
      "commands": [
        "webcompiler"
      ]
    }
  }
}
```

Now Heroku installs the tools needed and the application deploys.

But now I'm having this error, both locally and at heroku:

```
{
  "type": "https://httpstatuses.com/404",
  "title": "Not Found",
  "status": 404,
  "traceId": "00-bf156dfa91ad0e4a80ee76531210a9fc-141fc2a948dec54d-00"
}
```

**Comment by ivanmonteiro** [comment]
@nicolas63 I just realized that the backend is at one port, and the blazor frontend is at a different port.
It might be possible to start two separate processes at the Procfie (https://github.com/jhipster/jhipster-dotnetcore/blob/main/generators/heroku/templates/Procfile.ejs). 
But Heroku does not allow exposing ports so either the backend or the frontend would be served... =/
Since you're more experienced with Blazor, maybe we could brainstorm a solution for this

**Comment by nicolas63** [comment]
> I managed to do that by creating a new tools manifest (.config/dotnet-tools.json) with:
> `dotnet new tool-manifest` and adding the tools needed:
> `dotnet tool install -g Microsoft.Web.LibraryManager.Cli`
> and
> `dotnet tool install Excubo.WebCompiler`.
> 
> This creates .config/dotnet-tools.json file with the following contents:

it's a really good idea thanks 😃

**Comment by nicolas63** [comment]
> 
> 
> @nicolas63 I just realized that the backend is at one port, and the blazor frontend is at a different port.
> It might be possible to start two separate processes at the Procfie (https://github.com/jhipster/jhipster-dotnetcore/blob/main/generators/heroku/templates/Procfile.ejs).
> But Heroku does not allow exposing ports so either the backend or the frontend would be served... =/
> Since you're more experienced with Blazor, maybe we could brainstorm a solution for this

Yes it's two ports but i think we can deploy this on two dynos like the database no ? 
If this do not works we can extract the blazor integration in an other issue

**Comment by ivanmonteiro** [comment]
> > @nicolas63 I just realized that the backend is at one port, and the blazor frontend is at a different port.
> > It might be possible to start two separate processes at the Procfie (https://github.com/jhipster/jhipster-dotnetcore/blob/main/generators/heroku/templates/Procfile.ejs).
> > But Heroku does not allow exposing ports so either the backend or the frontend would be served... =/
> > Since you're more experienced with Blazor, maybe we could brainstorm a solution for this
> 
> Yes it's two ports but i think we can deploy this on two dynos like the database no ?
> If this do not works we can extract the blazor integration in an other issue

Actually the database is an addon. It "belongs" to the application/dyno.

But yes, with two separate apps maybe it could work:
- one for the backend with the db addon
- and another for the blazor frontend

From the [jincod's dotnetcore-buildpack docs](https://github.com/jincod/dotnetcore-buildpack):
`If repository contains multiple Web Applications (multiple Startup.* or Program.*), PROJECT_FILE and PROJECT_NAME environment variables allow to choose project for publishing.`

Each app would need a special PROJECT_FILE environment variable.

There is one small change needed on the generator for it to work:
The backend/api project "AssemblyName / executable name" is dasherized https://github.com/jhipster/jhipster-dotnetcore/blob/main/generators/server/templates/dotnetcore/src/Project/Project.csproj.ejs
But the convention by Heroku (and .Net in general) is to have the same "AssemblyName" as the project and its folder name. 
I'm not sure if there is a reason why It is setup as a dasherized name.

Example:
Project name: JhipsterSampleApp (PascalCased)
Project folder: JhipsterSampleApp (PascalCased)
Project csproj: JhipsterSampleApp.csproj (PascalCased)
**(current) AssemblyName / Executable name: jhipster-sample-app** (dasherized/kebab cased)
**(heroku convention) JhipsterSampleApp** (PascalCased)

So that would need to change for the application to work. Please let me know if you have any reasons to keep it dasherized.

**Comment by ivanmonteiro** [comment]
Check https://github.com/jincod/dotnetcore-buildpack/blob/de0a0876c095965906659248e70cea54b0315591/bin/compile#L77 for why it needs to change. Heroku extracts the executable from the PROJECT_NAME environment variable. Usually the project name has the same casing as the AssemblyName/executable. But we use dasherizedBaseName at the backend/api csproj instead of the usual pascalizedBaseName.

**Comment by nicolas63** [comment]
> Check https://github.com/jincod/dotnetcore-buildpack/blob/de0a0876c095965906659248e70cea54b0315591/bin/compile#L77 for why it needs to change. Heroku extracts the executable from the PROJECT_NAME environment variable. Usually the project name has the same casing as the AssemblyName/executable. But we use dasherizedBaseName at the backend/api csproj instead of the usual pascalizedBaseName. Heroku extracts it from eiter the first project it finds or environment variable if set.

Ok so i think i can change the assembly name i don't know why we use dasherized. I agree with heroku convention

**Comment by ivanmonteiro** [comment]
> > Check https://github.com/jincod/dotnetcore-buildpack/blob/de0a0876c095965906659248e70cea54b0315591/bin/compile#L77 for why it needs to change. Heroku extracts the executable from the PROJECT_NAME environment variable. Usually the project name has the same casing as the AssemblyName/executable. But we use dasherizedBaseName at the backend/api csproj instead of the usual pascalizedBaseName. Heroku extracts it from eiter the first project it finds or environment variable if set.
> 
> Ok so i think i can change the assembly name i don't know why we use dasherized. I agree with heroku convention

Don't worry, I can change it ASAP. I also think the heroku convention is better.

**Comment by ivanmonteiro** [comment]
The current status is:
Working:
- Monolithic deploys.

Databases available:
- mysql 
- postgres
- MS SQL Server (not tested because it is paid but It should work)

Not working:
- Blazor deploy is **not** working (I might neeed a little help from @nicolas63 to make it work)

Not supported:
- Microservice deploy is not supported. 
As far as I know it is difficult to run Consul on Heroku because of the dyno abstraction and lack of support to configure and open ports. Currently on Heroku I'm not even sure if Consul would work on heroku, but I think It would require hacks, which I'm not too fond of.
